---
title: "ggtree basics"
output: html_document
---

### [Install ggtree link](https://bioconductor.org/packages/release/bioc/html/ggtree.html#:~:text='ggtree'%20extends%20the%20'ggplot2,structures%20with%20their%20annotation%20data.)

See https://github.com/GaryNapier/ggtree for this markdown code and data

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

rm(list = ls())

setwd("~/Documents/ggtree/")

```

### Libraries:

```{r libraries, echo=T, out.width = "1500px", warning = F, message = F}

library(ggplot2) # ggtree is based on ggplot and uses a lot of its code
library(ggtree) # ggtree
library(treeio) # Loads up Beast trees
library(ggnewscale) # For adding new heatmaps (annotations)
library(scales) # For colours
library(gplots) # Also for colours
library(phytools) # Midpoint root
library(dplyr)
library(tidyr)

```


```{r functions, echo=FALSE, out.width = "1500px", warning = F, message = F}

# Functions ----

log10_ceiling <- function(x) {
  10^(ceiling(log10(x)))
}

```

```{r paths, echo=FALSE, out.width = "1500px", warning = F, message = F}

# Paths ----

metadata_path <- "metadata/"
plots_path <- "plots/"
beast_results_path <- "beast_results/"
newick_path <- "newick/"
gwas_results_path <- "gwas_results/"

```

```{r files, echo=FALSE, out.width = "1500px", warning = F, message = F}

# Files ----

# Metadata
metadata_file <- paste0(metadata_path, "pakistan_metadata.csv")
# Beast
mcc_tree_file <- paste0(beast_results_path, "PAKISTAN_ALL.mcc.tree")
beast_clusters_file <- paste0(metadata_path, "PAKISTAN_ALL.clusters.csv")
# Normal trees
tree_all_samps_newick_file <- paste0(newick_path, "PAKISTAN_ALL.filt.val.gt.g.snps.fa.treefile")
# GWAS
gwas_results_file <- paste0(gwas_results_path, "gwas_samples_mutations.txt")
mutations_tree_file <- paste0(plots_path, "mutations_tree.png")

```

### Read in files

```{r read-in-files, echo=T, out.width = "1500px", warning = F, message = F}

# Read in beast tree with treeio package
mcc_tree <- treeio::read.beast(mcc_tree_file)
metadata <- read.csv(metadata_file)
beast_clusters <- read.csv(beast_clusters_file, header = F, col.names = c("cluster", "id"))
mutations <- read.delim(gwas_results_file, header = T)
tree_all_samps <- read.newick(tree_all_samps_newick_file)

```

### Have a look at the Beast file

```{r , echo=T, out.width = "1500px", warning = F, message = F}

mcc_tree

```

There's lots of metadata here that can be used in the trees (the ```"with the following features available:"``` bit)


### Set up tree metadata - first and last dates etc

```{r , echo=T, out.width = "1500px", warning = F, message = F}

# I think this is how you get the first (estimated) date of the Beast tree - i.e. the root. 
# My root year happens to be 0, but if the root date was different, then I suspect the root.edge would be different 
# (i.e. root.edge is not actually in years)
first_date <- mcc_tree@phylo$root.edge
# Get the last overall date from the sample names - year after the last underscore
last_date <- as.numeric(max(unlist(lapply(strsplit(mcc_tree@phylo$tip.label, "_"), 
                                             function(x){x[length(x)]}))))
# In ggtree you have to give the last date as a character string in the form of YYYY-MM-DD
last_date_chr <- paste0(as.character(last_date), "-12-31")
# Total n samples
n_samps <- length(mcc_tree@phylo$tip.label)

```

### Trees 

Set up params/values for the trees:

```{r , echo=T, out.width = "1500px", warning = F, message = F}

# Expand the x-axis so there's a bit of room at the start and a very large space (80% of the full timeline) 
# to the right for the annotations
x_lim <- c(first_date-10, last_date+(last_date*0.8))
# Expand the y-axis so there's room at the top and bottom
y_lim <- c(-5, n_samps + (n_samps * 0.1))
# Define vertical line colour
v_line_col <- "red"
# Define sequence of year ticks on the x-axis
x_labs_full <- seq(first_date, last_date, by = 100)
# Angle of the labels so they fit in
angle <- 45

```

### Draw basic tree
```{r , echo=T, out.width = "1500px", warning = F, message = F}
ggtree(mcc_tree, mrsd = last_date_chr) 
```
  
### Add date scale
```{r , echo=T, out.width = "1500px", warning = F, message = F}
ggtree(mcc_tree, mrsd = last_date_chr) +
  theme_tree2() 
```


### Add annotations 
* 'confidence intervals' (red bars)
* posterior values of the branches ("1" etc.)
* tidy up the x-axis

Code for the tip labels is shown, but commented

```{r , echo=T, out.width = "1500px", warning = F, message = F}

ggtree(mcc_tree, mrsd = last_date_chr) + 
  theme_tree2() +
  # Tip labels
  # geom_tiplab(align=TRUE, linetype='dashed', linesize=.3, size = 2) +
  # Red confidence bars
  geom_range("length_0.95_HPD", color='red', size=2, alpha=.5) +
  # Text labels for branch posterior scores
  # Note - you can filter the values - here only 0.9 and above are shown.
  # Also can control the position with the vjust arg. 
  geom_text2(aes(label=round(as.numeric(posterior), 2),
                 subset=as.numeric(posterior)> 0.9,
                 x=branch), vjust=0) +
  # Set the y-axis as defined above
  coord_cartesian(ylim = y_lim) +
  # Set the x-axis limits, and year tick labels (don't know why you need 'breaks' as well as 'labels')
  scale_x_continuous(limits = x_lim, breaks = x_labs_full, labels = x_labs_full)+
  # Set the appearance of the year labels, including angle
  theme(axis.text.x = element_text(face = "bold", size = 8, angle = 45, hjust = 1))

```

### Add the metadata annotations to the big space on the right

#### Setup 

#### Prep data for heatmaps (strips)

```{r , echo=T, out.width = "1500px", warning = F, message = F}

# Add id/year hybrid col so samples names are correct - i.e. they match the Beast tree sample names
metadata$id_year <- paste0(metadata$wgs_id, "_", metadata$year)
rownames(metadata) <- metadata$id_year

# Subset metadata 
# FORMAT: one col per attribute, rownames of dataframe match names of samples in the tree
lin_data <- metadata[, "main_lineage", drop = F]
dr_status_data <- metadata[, "dr_status", drop = F]

# Convert lin and DR data to factors
lin_data <- data.frame(apply(lin_data, 2, as.factor))

# Change column names for neatness on tree
colnames(dr_status_data) <- "DR status"
colnames(lin_data) <- "Lineage"

```

Drug resistance data now looks like this. 
Note - the sample names are the **rownames** of the one-column dataframe.
Also note - some sample names are "XXXX_NA". This is because they've been pulled from the metadata and not all samples in the metadata have years assigned to them and therefore weren't used to create the tree. ggtree will find the ones that are in both the metadata and the tree, so it doesn't matter if you have extra samples in your metadata.
```{r , echo=T, out.width = "1500px", warning = F, message = F}
head(dr_status_data)
```

Lineage data is very similar:
```{r , echo=T, out.width = "1500px", warning = F, message = F}
head(lin_data)
```


#### Colours

```{r , echo=T, out.width = "1500px", warning = F, message = F}

# Define cols for each dataset - one colour per unique value of the datasets
alpha <- 0.9
lin_colours <- rainbow(length(unique(metadata$main_lineage)), alpha = alpha)
dr_status_colours <- scales::alpha(gplots::col2hex(c("green1", "yellow2", "orange1", "red1", "black", "grey")), alpha = alpha)

# Add names (unique values for each dataset) to dataframes. 
names(lin_colours) <- c(sort(unique(metadata$main_lineage)))
names(dr_status_colours) <- c("Sensitive", "Pre-MDR", "MDR", "Pre-XDR", "XDR", "Other")

```

Colour vectors:
```{r , echo=T, out.width = "1500px", warning = F, message = F}

lin_colours
dr_status_colours

```

Parameters for the strip annotations

```{r , echo=T, out.width = "1500px", warning = F, message = F}
# Where the annotations are placed (in this case 50 years to the right of the last date)
offset <- 50
# Width of the heatmap strip
width <- 0.05
```

To add the heatmap/strip annotation, draw the same tree as above and save as an object, then pass to ```gheatmap()```

```{r , echo=T, out.width = "1500px", warning = F, message = F}
gg_mcmc_tree <- ggtree(mcc_tree, mrsd = last_date_chr)+
  theme_tree2() +
  # geom_tiplab(align=TRUE, linetype='dashed', linesize=.3, size = 2) + 
  geom_range("length_0.95_HPD", color='red', size=2, alpha=.5) +
  geom_text2(aes(label=round(as.numeric(posterior), 2),
                 subset=as.numeric(posterior)> 0.9,
                 x=branch), vjust=0) +
  coord_cartesian(ylim = y_lim) +
  scale_x_continuous(breaks = x_labs_full, labels = x_labs_full, limits = x_lim)+
  theme(axis.text.x = element_text(face = "bold", size = 8, angle = 45, hjust = 1))

# Add lineage data 
lin_hm <- gheatmap(gg_mcmc_tree, lin_data,
                   width = width,
                   offset = offset, 
                   # Not sure what the color arg does or why it has to be NULL. Rest of the args are self-explanatory
                   color = NULL,
                   colnames_position = "top",
                   colnames_angle = angle, 
                   colnames_offset_y = 1,
                   hjust = 0,
                   font.size = 3) +
  # Add the custom colours defined above
  scale_fill_manual(values = lin_colours, breaks = names(lin_colours) )+
  # Define the legend title
  labs(fill = "Lineage")

lin_hm


```

Note - the ```scale_fill_manual()``` function took ages to work out - the ```values``` are the colours themseves, i.e. `r unname(lin_colours)`, and the ```breaks``` is the unique values in the data (i.e. `r names(lin_colours)`), which we've used to name the colour vector.

#### Add another strip (heatmap) - DR data

To do this we have to add the saved tree to the ```new_scale_fill()``` function from the ```ggnewscale``` package
# See "7.3.1 Visualize tree with multiple associated matrix" - https://yulab-smu.top/treedata-book/chapter7.html

It is possible to just have a dataset with two columns (i.e. in this case the lineage and DR data) and to concattenate the colours together, but this would mean there's only one legend and the whole thing is less clear. 

```{r , echo=T, out.width = "1500px", warning = F, message = F}

lin_hm <- lin_hm + ggnewscale::new_scale_fill() 

```

Note - ```offset``` has to be increased otherwise the strip will be drawn over the top of the previous one. 
The rest of the code is the same except for the tree input, data input, colours and legend label

```{r , echo=T, out.width = "1500px", warning = F, message = F}

dr_status_hm <- gheatmap(lin_hm, dr_status_data,
                         width = width,
                         # Increase offset
                         offset = offset+100, 
                         color = NULL,
                         colnames_position = "top",
                         colnames_angle = angle, 
                         colnames_offset_y = 1,
                         hjust = 0,
                         font.size = 3) +
  # Add colours for DR status
  scale_fill_manual(values = dr_status_colours, breaks = names(dr_status_colours) )+
  # New legend label
  labs(fill = "DR\nstatus")

dr_status_hm

```

#### Add binary dataset

The last annotation to add is a binary dataset of individual drugs - 0/1 whether the sample is resistant to the drug

Get data and convert numeric 0/1 to factors

```{r , echo=T, out.width = "1500px", warning = F, message = F}
dr_data <- dplyr::select(metadata, rifampicin:delamanid)
dr_data <- data.frame(apply(dr_data, 2, as.factor))
```
Lots of columns with 0/1/NA
```{r , echo=T, out.width = "1500px", warning = F, message = F}
head(dr_data)
```

Add the tree to ```new_scale_fill()``` again:
```{r , echo=T, out.width = "1500px", warning = F, message = F}

dr_status_hm <- dr_status_hm + ggnewscale::new_scale_fill() 

```

This time ```color``` is changed to ```"black"``` I think as the 'base' colour, and ```low``` and ```high``` added. I think these assign the 0 value to white and the 1 to black. NA is defined in the ```scale_fill_manual()``` function.

This time in ```scale_fill_manual()```, instead of ```breaks```, for some reason ```labels``` is used to map the 0/1 to the legend labels.

```{r , echo=T, out.width = "1500px", warning = F, message = F}

gheatmap(dr_status_hm, dr_data,
         # Increase offset
         offset = offset+200,
         width = width+0.5,
         # Change color to black
         # color = NULL,
         color="black",
         low="white", 
         high="black", 
         colnames_position = "top",
         colnames_angle = angle, 
         colnames_offset_y = 1,
         hjust = 0,
         font.size = 2.5) +
  # Define colours
  scale_fill_manual(values=c("white", "black"), labels = c("Sensitive", "Resistant", "NA"), na.value = "grey")+
  labs(fill = "Drug\nresistance")

```

The drug binary data looks a bit squashed on this webpage but I'll show how to save to pdf later.

Actually the whole tree is a bit squashed at the tips because the samples were taken from possible transmission cases, but are also from different lineages so have MRCA a long time ago. So it might be useful to zoom in on the tips.

#### Zoom in to the tree

If we specify a year to limit the x-axis, then only those clades that have MRCA after this year are drawn, which might be useful for zooming in on clusters. 


Zoom tree setup

The code is basically the same, but the x-axis limits are changed. Note - the ```offset``` and ```width``` have to be much smaller.

I've also added a vertical red line to show when the last date was.

```{r , echo=T, out.width = "1500px", warning = F, message = F}

zoom_range <- c((last_date+1) - 50, (last_date + 1) + 50)
zoom_range_seq <- seq(zoom_range[1], (last_date+1), 2)

offset_zoom <- 0
width_zoom <- 0.001

```

```{r , echo=T, out.width = "1500px", warning = F, message = F}

gg_mcmc_tree_zoom <- ggtree(mcc_tree, mrsd = last_date_chr) +
  theme_tree2() +
  coord_cartesian(ylim = y_lim) +
  scale_x_continuous(breaks = zoom_range_seq,
                     labels = zoom_range_seq,
                     limits = zoom_range) +
  # Add vertical line
  geom_vline(aes(xintercept = 2020), col = "red")+
  theme(axis.text.x = element_text(face = "bold", size = 8, angle = 45, hjust = 1))

# Add lin data 
lin_hm <- gheatmap(gg_mcmc_tree_zoom, lin_data,
                   width = width_zoom,
                   offset = offset_zoom, 
                   color = NULL,
                   colnames_position = "top",
                   colnames_angle = angle, colnames_offset_y = 1,
                   hjust = 0,
                   font.size = 3) +
  scale_fill_manual(values = lin_colours, breaks = names(lin_colours) )+
  labs(fill = "Lineage")

lin_hm <- lin_hm + ggnewscale::new_scale_fill() 

# Add DR status
dr_status_hm <- gheatmap(lin_hm, dr_status_data,
                         width = width_zoom,
                         offset = offset_zoom + 2, 
                         color = NULL,
                         colnames_position = "top",
                         colnames_angle = angle, colnames_offset_y = 1,
                         hjust = 0,
                         font.size = 3) +
  scale_fill_manual(values = dr_status_colours, breaks = names(dr_status_colours) )+
  labs(fill = "DR\nstatus")

# Add DR individual status
dr_status_hm <- dr_status_hm + ggnewscale::new_scale_fill() 

gheatmap(dr_status_hm, dr_data,
         offset = offset_zoom + 4,
         width = width_zoom + 0.02,
         # color = NULL,
         low="white", high="black", color="black",
         colnames_position = "top",
         colnames_angle = angle, 
         colnames_offset_y = 1,
         hjust = 0,
         font.size = 2.5, 
         legend_title = "llw") +
  scale_fill_manual(values=c("white", "black"), labels = c("Sensitive", "Resistant", "NA"), na.value = "grey")+
  labs(fill = "Drug\nresistance")


```


#### Loop over data to plot individual clusters

If we want to plot these individual clusters it's possible to loop over lists of samples, subset the tree, and run the ggtree code

This is the list of samples and their groups - all the clusters that result from cutting the tree at 50 years before the last sample. This file was created with my [cut_tree.py](https://github.com/GaryNapier/pakistan/tree/main/python_scripts) function.

```{r , echo=T, out.width = "1500px", warning = F, message = F}

head(beast_clusters)

```

Split the file into each cluster to loop over:
```{r , echo=T, out.width = "1500px", warning = F, message = F}

clusters_split <- split(beast_clusters, beast_clusters$cluster)
head(clusters_split)

```

Notes:

* Usually I'd save these as a pdf (one plot per page, and one page per loop) - the relevent code is  ```pdf(file = <filename>)``` at the start of the loop to open the pdf file, ```print(<plot_name>)``` during the loop, and ```dev.off()``` at the end to finish saving it. These lines are commented here, except the ```print(<plot_name>)``` so I can show the plots here. 

* There's only one function in treeio to subset trees, which is ```treeio::drop.tip```, so I have to first get the samples that *aren't* in the cluster of interest with ```%in%```

* I use a new way here of getting first and last dates by creating the basic tree first - the ```theme_tree2()``` obviously does some calculations on the edge lengths to get the dates, but I haven't pulled out the function, rather I just run it and get the dates. 

* Here I program the x-axis scale and labels according to the order of magnitude (oom) of the date range. If I just leave it to the defaults it's a bit a of a mess. The ```log10_ceiling()``` function determines the oom.

* I also have to determine the x-axis ranges and the space to the right of the tree for the annotations by taking proportions of the date range - so if a cluster's MRCA stretches back 50 years compared to 5 years, the spaces can get too squashed or stretched, so evserything's done as percentages of the time span. This also applies to the offsets and widths, which are dependent on the x-axis.

* Similarly, I also have to change the y-axis space because the number of samples in each cluster changes a lot. 

```{r , echo=T, out.width = "1500px", warning = F, message = F}

# Get sample names from whole tree
mcc_tree_tips <- mcc_tree@phylo$tip.label

# pdf(file = beast_clusters_pdf_file)
for(i in seq(clusters_split)){
  
  # Get the names of samples from cluster i
  names_to_keep <- clusters_split[[i]]$id
  # Get the names from the full Beast tree that aren't these names
  names_to_drop <- mcc_tree_tips[!(mcc_tree_tips %in% names_to_keep)]
  # Create new sub tree 
  clust_tree <- treeio::drop.tip(mcc_tree, names_to_drop)
  
  # Do tree first to get first and last date and ranges
  last_date_clust <- as.numeric(max(unlist(lapply(strsplit(clust_tree@phylo$tip.label, "_"), function(x){x[length(x)]}))))
  last_date_clust_chr <- paste0(last_date_clust-1, "-12-31")
  gg_clust_tree <- ggtree(clust_tree, mrsd = last_date_clust_chr) +
    theme_tree2()
  last_date_clust <- ceiling(max(gg_clust_tree$data$x))
  first_date_clust <- floor(min(gg_clust_tree$data$x))
  date_range_clust <- c(first_date_clust, last_date_clust)
  year_span_clust <- diff(date_range_clust)
  
  # N samples in cluster i
  n_samps <- length(clust_tree@phylo$tip.label)
  
  # Get order of magnitude of the date range
  oom <- log10_ceiling(year_span_clust)
  if(oom == 1){
    by <- 1
  }else if(oom == 10){
    by <- 2
  }else if(oom == 100){
    by <- 10
  }else{
    by <- 100
  }
  # Create seq of dates for x-axis
  zoom_range_seq_clust <- seq(first_date_clust,
                              last_date_clust,
                              by = by)
  
  # Determine space to right of tree for annotations as a multiple of total time range
  post_tree_multiply <- 5
  post_tree_span <- (year_span_clust * post_tree_multiply)
  x_lim_clust <- c(first_date_clust - (year_span_clust*0.2), last_date_clust + post_tree_span)
  # Same for y-axis according to number of samples
  y_lim_clust <- c(0 - floor(n_samps*0.1),  n_samps + ceiling(n_samps*0.2) )
  # Sort out the offsets and widths
  width_clust <- 0.035 * post_tree_multiply  # % of tree time span
  offset_clust <- (width_clust * year_span_clust)
  
  # Make tree
  gg_clust_tree <- ggtree(clust_tree, mrsd = last_date_clust_chr) %<+% metadata +
    theme_tree2() +
    coord_cartesian(ylim = y_lim_clust) +
    scale_x_continuous(breaks = zoom_range_seq_clust,
                       labels = zoom_range_seq_clust,
                       limits = x_lim_clust) +
    geom_vline(aes(xintercept = last_date_clust), col = "red")+
    theme(axis.text.x = element_text(face = "bold", size = 8, angle = 45, hjust = 1))
  
  # Add lin data
  lin_hm_clust <- gheatmap(gg_clust_tree, lin_data,
                           width = width_clust,
                           offset = 0,
                           color = NULL,
                           colnames_position = "top",
                           colnames_angle = angle,
                           # colnames_offset_y = 1,
                           hjust = 0,
                           font.size = 3) +
    scale_fill_manual(values = lin_colours, breaks = names(lin_colours) )+
    labs(fill = "Lineage")
  
  lin_hm_clust <- lin_hm_clust + ggnewscale::new_scale_fill()
  
  # Add DR status clusters
  dr_status_hm_clust <- gheatmap(lin_hm_clust, dr_status_data,
                                 width = width_clust,
                                 offset = offset_clust,
                                 color = NULL,
                                 colnames_position = "top",
                                 colnames_angle = angle,
                                 # colnames_offset_y = 1,
                                 hjust = 0,
                                 font.size = 3) +
    scale_fill_manual(values = dr_status_colours, breaks = names(dr_status_colours) )+
    labs(fill = "DR\nstatus")
  
  dr_status_hm_clust <- dr_status_hm_clust + ggnewscale::new_scale_fill()
  
  final_plot <- gheatmap(dr_status_hm_clust, dr_data,
                         width = width_clust * ncol(dr_data),
                         # offset = ceiling(offset_clust * 3),
                         offset = (offset_clust * 2) + (offset_clust * 0.2),
                         # color = NULL,
                         low="white", high="black", color="black",
                         colnames_position = "top",
                         colnames_angle = angle,
                         colnames_offset_y = 0,
                         hjust = 0,
                         font.size = 2.5,
                         legend_title = "llw") +
    scale_fill_manual(values=c("white", "black"), labels = c("Sensitive", "Resistant", "NA"), na.value = "grey")+
    labs(fill = "Drug\nresistance")
  
  print(final_plot)
  
}
# dev.off()


```


#### Advanced

Say we want to see where some mutations lie on a tree. These mutations are within a certain number of genes, so we can just do one heatmap strip per gene, as we've done with lineage and drug resistance above.

However, some samples have more than one mutation per gene. We could make a new strip, one for each mutation, but the data are rather sparse and we want it to be clear that the mutations are within their respective genes. 

Here's a look at the mutations data:

```{r gwas-mutations-data, echo=T, out.width = "1500px", warning = F, message = F}

head(mutations)

```

There's one sample per mutation, and that mutation belongs to a gene

Hoever, some samples have more than one mutation per gene:

```{r gwas-mutations-data-overlaps, echo=T, out.width = "1500px", warning = F, message = F}

x <- mutations[mutations["gene"] == "proX", ]

x <- x[duplicated(x$wgs_id) | duplicated(x$wgs_id, fromLast = T), ]

head(x[order(x$wgs_id),])

```

Here for example ERR2510252 has two mutations on proX gene. 

To represent this we can split each heatmap strip into sub-strips for each gene. 

Here is the final plot, just to make clear what it will look like. 

The proX and Rv2307B genes have been split into sub-strips.

```{r mutations-tree-png, echo=T, out.width = "1500px", warning = F, message = F}

knitr::include_graphics(mutations_tree_file)

```

First we have to wrangle the data to get it into the right format. 

The genes that do (proX and Rv2307B) and don't (murA, nusG, wag31) have overlaps will have to be wrangled differently. So it's best to split the data by genes, make a list, and loop over the list, applying different rules to the two types of data.

Now we have a dataframe of each gene in a list:

```{r mutations-wrangling-split, echo=T, out.width = "1500px", warning = F, message = F}

# Split on gene
mutations_split <- split(mutations, mutations$gene)

lapply(mutations_split, head)

```

While the data in in this format, the colours can easily be defined, so do this now.

```{r mutations-colours, echo=T, out.width = "1500px", warning = F, message = F}

# ------------------
# | Define colours |
# ------------------
mutations_cols <- lapply(mutations_split, function(x){
  rainbow(length(unique(x$mutation)))
})

# Make the NAs white and assign names
# n.b. this is done in a loop rather than lapply(). For some reason it can't be easily done in lapply()
for(i in seq(mutations_cols)){
  mutations_cols[[i]][length(mutations_cols[[i]])+1] <- "#FFFFFF"
}

for(i in seq(mutations_cols)){
  names(mutations_cols[[i]]) <- c(unique(mutations_split[[i]]$mutation), NA)
}

```

...now continue the wrangling...

```{r gwas-genes-mutations-tree, echo=T, out.width = "1500px", warning = F, message = F}

# ----------------------
# | Continue wrangling |
# ----------------------

# Get which lists have overlappling mutations/samples
which_in_list_overlap <- vector()
for(i in seq(mutations_split)){
  if(any(duplicated(mutations_split[[i]]$wgs_id))){
    which_in_list_overlap[i] <- i
  }
}
# Remove NA
which_in_list_overlap <- which_in_list_overlap[!(is.na(which_in_list_overlap))]

which_in_list_overlap

# Make a pivot of the dataframes where there are samples with more than one mutation,
# where each col is a mutation - if the sample has the mutation, value is the mutation and NA if the
# sample does not have it.
for(i in seq(mutations_split)){
  if(i %in% which_in_list_overlap){
    mutations_split[[i]] <- reshape2::dcast(mutations_split[[i]], wgs_id ~ mutation, value.var = "mutation")
  }
}

lapply(mutations_split, head)

# Clean up/wrangle
for(i in seq(mutations_split)){
  # Merge in the rest of the samples in the tree - all of the rest will be NA for the gene and mutation
  mutations_split[[i]] <- merge(metadata["wgs_id"], mutations_split[[i]],
                                by = "wgs_id", all.x = T, sort = F)
  # Convert the rownames the IDs to match to the tree, as with the lineage and DR data above
  row.names(mutations_split[[i]]) <- mutations_split[[i]]$wgs_id
  # Drop the ID column and the gene column
  mutations_split[[i]] <- mutations_split[[i]][!(names(mutations_split[[i]]) %in% c("wgs_id", "gene"))]
  # Change the column names of the mutations column from 'mutation' to the name of the gene (names(mutations_split[i]))
  # If the dataframe is one of the 'overlap' genes, then the columns have to be named uniquely or ggtree throws and error
  if(i %in% which_in_list_overlap){
    colnames(mutations_split[[i]]) <- paste0(names(mutations_split[i]), "_", 1:ncol(mutations_split[[i]]))
  }else{
    colnames(mutations_split[[i]]) <- names(mutations_split[i])
  }
}

lapply(mutations_split, head)


```

Now plot the data as heatmap strips.

For the 'overlap' genes (proX and Rv2307B) - because we have one column for each mutation, the 'substrips' will automatically be created. 

```{r gwas-genes-mutations-tree, echo=T, out.width = "1500px", warning = F, message = F}

# -----
# PLOT
# -----

# Setup

tree_all_samps <- phytools::midpoint.root(tree_all_samps)
n_samps <- length(tree_all_samps$tip.label)
width <- 0.05
font_sz <- 3
line_sz <- 0.25
angle <- 45
y_lim <- c(-10, n_samps + (n_samps * 0.1))
legend_spec <- theme(legend.title = element_text(size = 9),
                     legend.text = element_text(size = 7),
                     legend.key.size = unit(0.3, "cm"))

# Make basic tree
ggtree_all_samps <- ggtree(tree_all_samps, size = line_sz) +
coord_cartesian(ylim = y_lim)


# Add the mutations data
offset <- 0
y_annotate <- n_samps+15
mutations_tree <- ggtree_all_samps # Rename so the loop works (otherwise will just keep plotting the first tree)
increase_offset_by <- 0.0014
for(i in seq(mutations_split)){
# for(i in 1:3){

  gene <- names(mutations_cols[i])
  # mutations_tree <- gheatmap(mutations_tree, mutations_split[[i]],
  #          width = width,
  #          offset = offset,
  #          color = NULL,
  #          colnames_position = "top",
  #          colnames_angle = angle,
  #          colnames_offset_y = 1,
  #          hjust = 0,
  #          font.size = 3,
  #          colnames = T) +
  #   # annotate("text", x = 0.027, y = y_annotate, size = font_sz, angle = angle, label = gene) +
  #   scale_fill_manual(values = mutations_cols[[i]], breaks = names(mutations_cols[[i]]) ) +
  #   # scale_x_discrete(expand = c(1,0)) +
  #   labs(fill = gene) +
  #   legend_spec
  
  pg <- ggplot_build(mutations_tree)
  pg <- pg$data[[(i * 2) - 1]]
  
  x_pos <- max(unique(pg$x))

  mutations_tree <- gheatmap_custom(mutations_tree, mutations_split[[i]], pg,
           width = width,
           offset = offset,
           color = NULL,
           colnames_position = "top",
           colnames_angle = angle,
           colnames_offset_y = 1,
           hjust = 0,
           font.size = 3,
           colnames = F) +
    annotate("text", x = x_pos, y = y_annotate, size = font_sz, angle = angle, label = gene) +
    scale_fill_manual(values = mutations_cols[[i]], breaks = names(mutations_cols[[i]]) ) +
    # scale_x_discrete(expand = c(1,0)) +
    labs(fill = gene) +
    legend_spec

  mutations_tree <- mutations_tree + ggnewscale::new_scale_fill()
  
  offset <- offset + increase_offset_by

}

mutations_tree





offset_start <- 0
y_annotate <- n_samps+12
increase_offset_by <- 0.0013
# for(i in seq(mutations_split)){
i <- 1
mutation <- names(mutations_cols[i])
mutations_tree <- ggtree_all_samps
mutations_tree <- gheatmap(mutations_tree, mutations_split[[i]],
         width = width,
         offset = offset_start,
         color = NULL,
         colnames_position = "top",
         colnames_angle = angle,
         colnames_offset_y = 1,
         hjust = 0,
         font.size = 3,
         colnames = F) +
  # annotate("text", x = 0.027, y = y_annotate, size = font_sz, angle = angle, label = mutation) +
  scale_fill_manual(values = mutations_cols[[i]], breaks = names(mutations_cols[[i]]) ) +
  labs(fill = mutation) +
  legend_spec

mutations_tree <- mutations_tree + ggnewscale::new_scale_fill()
 # offset <- offset+0.0035
# }

offset <- offset + increase_offset_by
i <- 2
mutation <- names(mutations_cols[i])
mutations_tree <- gheatmap(mutations_tree, mutations_split[[i]],
         width = width,
         offset = offset,
         color = NULL,
         colnames_position = "top",
         colnames_angle = angle,
         colnames_offset_y = 1,
         hjust = 0,
         font.size = 3,
         colnames = F) +
  # annotate("text", x = 0.0285, y = y_annotate, size = font_sz, angle = angle, label = mutation) +
  scale_fill_manual(values = mutations_cols[[i]], breaks = names(mutations_cols[[i]]) ) +
  labs(fill = mutation) +
  legend_spec

mutations_tree <- mutations_tree + ggnewscale::new_scale_fill()

# offset <- offset+offset_start+0.0005
# i <- 3
# mutation <- names(mutations_cols[i])
# trans_tree <- gheatmap(trans_tree, mutations_split[[i]],
#          width = width,
#          offset = offset,
#          color = NULL,
#          colnames_position = "top",
#          colnames_angle = angle,
#          colnames_offset_y = 1,
#          hjust = 0,
#          font.size = 3,
#          colnames = F) +
#   annotate("text", x = 0.0295, y = y_annotate, size = font_sz, angle = angle, label = mutation) +
#   scale_fill_manual(values = mutations_cols[[i]], breaks = names(mutations_cols[[i]]) ) +
#   labs(fill = mutation) +
#   legend_spec
# 
# trans_tree <- trans_tree + ggnewscale::new_scale_fill()
# 
# offset <- offset+offset_start
# i <- 4
# mutation <- names(mutations_cols[i])
# trans_tree <- gheatmap(trans_tree, mutations_split[[i]],
#          width = width,
#          offset = offset,
#          color = NULL,
#          colnames_position = "top",
#          colnames_angle = angle,
#          colnames_offset_y = 1,
#          hjust = 0,
#          font.size = 3,
#          colnames = F) +
#   annotate("text", x = 0.0312, y = y_annotate+6, size = font_sz, angle = angle, label = mutation) +
#   scale_fill_manual(values = mutations_cols[[i]], breaks = names(mutations_cols[[i]]) ) +
#   labs(fill = mutation) +
#   legend_spec
# 
# trans_tree <- trans_tree + ggnewscale::new_scale_fill()
# 
# offset <- offset+offset_start-0.0005
# i <- 5
# mutation <- names(mutations_cols[i])
# trans_tree <- gheatmap(trans_tree, mutations_split[[i]],
#          width = width,
#          offset = offset,
#          color = NULL,
#          colnames_position = "top",
#          colnames_angle = angle,
#          colnames_offset_y = 1,
#          hjust = 0,
#          font.size = 3,
#          colnames = F) +
#   annotate("text", x = 0.0323, y = y_annotate, size = font_sz, angle = angle, label = mutation) +
#   scale_fill_manual(values = mutations_cols[[i]], breaks = names(mutations_cols[[i]]) ) +
#   labs(fill = mutation) +
#   legend_spec
# 
# trans_tree <- trans_tree + ggnewscale::new_scale_fill()
# 
# trans_tree

png(filename = "../plots/mutations_tree.png", width = 2500, height = 2000, res = 300)
trans_tree
dev.off()

```








